plenum/common/timer.py

```diff
@@ -0,0 +1,76 @@
+from abc import ABC, abstractmethod
+from bisect import bisect_right
+from typing import Callable
+
+import time
+
+
+class TimerInterface(ABC):
+    @abstractmethod
+    def schedule(self, delay: int, callback: Callable):
+        pass
+
+    @abstractmethod
+    def cancel(self, callback: Callable):
+        pass
+
+
+class Timer(TimerInterface):
+    def __init__(self, get_current_time=time.perf_counter):
+        self._get_current_time = get_current_time
+        self._timestamps = []
```

Comment:
In fact first attempt at implementing this actually used `SortedSet`, however it failed test when same callback was scheduled for same time twice (obviously it was same item, so `SortedSet` discarded one of them). I'm not sure if it is actually desirable behavior or not. Speaking of `SortedDict` - it won't fit here, since there could be multiple callbacks per timestamp, but it will keep only one value per key. And while writing all this I checked docs on sorted containers and found that there is also `SortedList` which looks like a good candidate here.

Reply:
What about SortedDict where the key is a timestamp and value is a list of callbacks?

Reply:
Didn't think about it, it's nice idea, but will still require more logic than `SortedList`.